File: mainActivity/src/main/java/com/sierrawireless/avphone/ConfigureFragment.java
package com.sierrawireless.avphone;

import net.airvantage.utils.AvPhonePrefs;
import net.airvantage.utils.PreferenceUtils;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.TextView;

import com.baoyz.swipemenulistview.SwipeMenu;
import com.baoyz.swipemenulistview.SwipeMenuCreator;
import com.baoyz.swipemenulistview.SwipeMenuItem;
import com.baoyz.swipemenulistview.SwipeMenuListView;
import com.sierrawireless.avphone.adapter.ObjectAdapter;
import com.sierrawireless.avphone.auth.AuthUtils;
import com.sierrawireless.avphone.auth.Authentication;
import com.sierrawireless.avphone.message.IMessageDisplayer;
import com.sierrawireless.avphone.model.AvPhoneObject;
import com.sierrawireless.avphone.task.DeleteSystemListener;
import com.sierrawireless.avphone.task.DeleteSystemResult;
import com.sierrawireless.avphone.task.DeleteSystemTask;
import com.sierrawireless.avphone.task.IAsyncTaskFactory;
import com.sierrawireless.avphone.task.SyncWithAvListener;
import com.sierrawireless.avphone.task.SyncWithAvParams;
import com.sierrawireless.avphone.task.SyncWithAvResult;
import com.sierrawireless.avphone.task.SyncWithAvTask;
import com.sierrawireless.avphone.tools.Tools;

import java.util.ArrayList;

public class ConfigureFragment extends AvPhoneFragment {
    private static final String TAG = "ConfigureFragment";

    private ObjectsManager objectsManager;
    ArrayList<String> menu;
    public static String INDEX = "index";
    public static int CONFIGURE = 0;
    public static String POS = "position";
    private boolean delete;

    private View view;


    private IAsyncTaskFactory taskFactory;

    public ConfigureFragment() {
        super();
    }

    public void setTaskFactory(IAsyncTaskFactory taskFactory) {
        this.taskFactory = taskFactory;
    }



    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {


        view = inflater.inflate(R.layout.fragment_configure, container, false);

        objectsManager = ObjectsManager.getInstance();


        menu = new ArrayList<>();

        for (AvPhoneObject object: objectsManager.objects) {
            menu.add(object.name);
        }

        SwipeMenuListView listView = view.findViewById(R.id.objectConfigure);

        SwipeMenuCreator creator = new SwipeMenuCreator() {

            @Override
            public void create(SwipeMenu menu) {

                switch (menu.getViewType()) {
                    case 0:
                        // create "delete" item
                        SwipeMenuItem deleteItem = new SwipeMenuItem(
                                getActivity().getBaseContext());
                        // set item background
                        deleteItem.setBackground(new ColorDrawable(Color.rgb(0xF9,
                                0x3F, 0x25)));
                        // set item width
                        deleteItem.setWidth((int) Tools.dp2px(
                                getActivity().getBaseContext()));
                        // set a icon
                        deleteItem.setIcon(android.R.drawable.ic_menu_delete);
                        // add to menu
                        menu.addMenuItem(deleteItem);
                        break;
                }
            }


        };


        listView.setOnMenuItemClickListener(new SwipeMenuListView.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(int position, SwipeMenu lmenu, int index) {
                switch (index) {
                    case 0:
                        //delete
                        objectsManager.setSavedPosition(position);
                        delete();
                        delete = true;
                        break;

                }
                return false;
            }
        });

        listView.setMenuCreator(creator);

        ObjectAdapter adapter = new ObjectAdapter(getActivity(), android.R.layout.simple_list_item_1, menu);

        listView.setAdapter(adapter);

        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                //Open a new intent with the selected Object
                Log.d(TAG, "onItemClick: " + i + " " +menu.get(i));
                Intent intent = new Intent(view.getContext(), ObjectConfigureActivity.class);
                intent.putExtra(INDEX, i);

                startActivityForResult(intent, CONFIGURE);

            }
        });


        Button doneBtn = view.findViewById(R.id.doneConfigureBtn);
        Button addBtn = view.findViewById(R.id.addConfigureBtn);

        doneBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ((MainActivity)getActivity()).goHomeFragment();
            }
        });

        addBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(view.getContext(), ObjectConfigureActivity.class);
                intent.putExtra(INDEX, -1);

                startActivityForResult(intent, CONFIGURE);
            }
        });


        return view;
    }




    private boolean checkCredentials() {

        AvPhonePrefs prefs = PreferenceUtils.getAvPhonePrefs(getActivity());

        if (!prefs.checkCredentials()) {
            PreferenceUtils.showMissingPrefsDialog(getActivity());
            return false;
        }

        return true;

    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == AuthorizationActivity.REQUEST_AUTHORIZATION) {
            Authentication auth = AuthUtils.activityResultAsAuthentication(requestCode, resultCode, data);
            if (auth != null) {
                authManager.onAuthentication(auth);
                if (delete) {
                    deleteSysten(auth.getAccessToken());
                }else {
                    syncWithAv(auth.getAccessToken());
                }
            }
        }else if (requestCode == CONFIGURE) {
            if (resultCode == Activity.RESULT_OK) {
                this.delete = false;
                int position = data.getIntExtra(POS, -1);
                //set current and start synchronization

                objectsManager.setSavedPosition(position);

                if (checkCredentials()) {
                    Authentication auth = authManager.getAuthentication();
                    if (auth != null && !auth.isExpired()) {
                        syncWithAv(auth.getAccessToken());
                    } else {
                        this.delete = false;
                        requestAuthentication();
                    }
                }
            }
            MainActivity.instance.loadMenu();


        }
    }

    protected void delete() {
        if (checkCredentials()) {
            Authentication auth = authManager.getAuthentication();
            if (auth != null && !auth.isExpired()) {
                deleteSysten(auth.getAccessToken());
            } else {
                this.delete = true;
                requestAuthentication();
            }
        }
    }

    private void deleteSysten(String token) {
        AvPhonePrefs prefs = PreferenceUtils.getAvPhonePrefs(getActivity());

        final IMessageDisplayer display = this;

        final DeleteSystemTask deleteTask = taskFactory.deleteSystemTak(prefs.serverHost, token);
        deleteTask.execute();

        deleteTask.addProgressListener(new DeleteSystemListener() {

            public void onDeleting(DeleteSystemResult result) {
                Log.d(TAG, "onSynced: ICI");
                if (delete) {
                    objectsManager.removeSavedObject();
                }

                deleteTask.showResult(result, display, getActivity());
                MainActivity.instance.loadMenu();


            }
        });

    }

    private void syncWithAv(String token) {

        AvPhonePrefs prefs = PreferenceUtils.getAvPhonePrefs(getActivity());

        final IMessageDisplayer display = this;

        final SyncWithAvTask syncTask = taskFactory.syncAvTask(prefs.serverHost, token);

        SyncWithAvParams syncParams = new SyncWithAvParams();
        syncParams.deviceId = DeviceInfo.getUniqueId(getActivity());
        syncParams.imei = DeviceInfo.getIMEI(getActivity());
        syncParams.deviceName = DeviceInfo.getDeviceName();
        syncParams.iccid = DeviceInfo.getICCID(getActivity());
        syncParams.mqttPassword = prefs.password;
        syncParams.customData = PreferenceUtils.getCustomDataLabels(getActivity());

        syncTask.execute(syncParams);

        syncTask.addProgressListener(new SyncWithAvListener() {
            @Override
            public void onSynced(SyncWithAvResult result) {
                Log.d(TAG, "onSynced: ICI");
                if (delete) {
                    objectsManager.removeSavedObject();
                }

                syncTask.showResult(result, display, getActivity());
                MainActivity.instance.loadMenu();

                if (!result.isError()) {
                    syncListener.onSynced(result);
                }

            }
        });

    }


    public TextView getErrorMessageView() {
       return (TextView) view.findViewById(R.id.configure_error_message);
   }

}

File: mainActivity/src/main/java/com/sierrawireless/avphone/ConfigureFragment.kt
package com.sierrawireless.avphone

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import com.crashlytics.android.Crashlytics
import com.sierrawireless.avphone.activity.AuthorizationActivity
import com.sierrawireless.avphone.activity.MainActivity
import com.sierrawireless.avphone.activity.ObjectConfigureActivity
import com.sierrawireless.avphone.adapter.ObjectAdapter
import com.sierrawireless.avphone.auth.AuthUtils
import com.sierrawireless.avphone.task.IAsyncTaskFactory
import com.sierrawireless.avphone.task.SyncWithAvParams
import com.sierrawireless.avphone.task.UpdateParams
import com.sierrawireless.avphone.tools.DeviceInfo
import kotlinx.android.synthetic.main.fragment_configure.*
import net.airvantage.utils.PreferenceUtils
import java.util.*

open class ConfigureFragment : AvPhoneFragment() {
    override var errorMessageView: TextView? = null

    private var objectsManager: ObjectsManager? = null
    private var menu: ArrayList<String> = ArrayList()
    enum class Mode {
        DELETE,
        SYNC,
        UPDATE

    }
    internal var delete: Mode = Mode.SYNC

    private var lView: View? = null

    var current = 0


    private var taskFactory: IAsyncTaskFactory? = null

    fun setTaskFactory(taskFactory: IAsyncTaskFactory) {
        this.taskFactory = taskFactory
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {


        lView = inflater.inflate(R.layout.fragment_configure, container, false)



        return lView
    }

    private fun reloadMenu(){
        menu = ArrayList()

        for (obj in objectsManager!!.objects) {
            menu.add(obj.name!!)
        }


        val adapter = ObjectAdapter(activity, R.layout.menu_objects, menu)
        objectConfigure.adapter = adapter
    }
    override fun onStart() {
        instance = this
        super.onStart()
        objectsManager = ObjectsManager.getInstance()
        errorMessageView = configure_error_message


       reloadMenu()

        objectConfigure.onItemClickListener = AdapterView.OnItemClickListener { _, view, i, _ ->
            val deleteActionBtn: Button = view.findViewById(R.id.menuDeleteActionBtn)
            if (deleteActionBtn.visibility == View.VISIBLE) {
                val deleteBtn:ImageButton = view.findViewById(R.id.menuDeleteBtn)
                deleteBtn.visibility = View.VISIBLE
                deleteActionBtn.visibility = View.GONE
            } else {
                startObjectConfigure(i)
            }
        }

        doneConfigureBtn.setOnClickListener { (activity as MainActivity).goLastFragment() }

        addConfigureBtn.setOnClickListener { _ ->
            startObjectConfigure(-1)
        }

        resync.setOnClickListener { _ ->
            resyncAll()
        }
    }

    private fun startObjectConfigure(position: Int) {

        //Open a new intent with the selected Object
        val intent = Intent(view?.context, ObjectConfigureActivity::class.java)
        intent.putExtra(INDEX, position)

        startActivityForResult(intent, CONFIGURE)
    }

    override fun onStop() {
        super.onStop()
        instance = null
    }

    private fun checkCredentials(): Boolean {

        val prefs = PreferenceUtils.getAvPhonePrefs(activity)

        if (!prefs.checkCredentials()) {
            PreferenceUtils.showMissingPrefsDialog(activity)
            return false
        }

        return true

    }

    private fun resyncAll() {
        current = 0
        // first check credential
        if (checkCredentials()) {
            val auth = authManager!!.authentication
            if (!auth!!.isExpired) {
                updateAllSystem(auth.accessToken!!)
            } else {
                this.delete = Mode.UPDATE
                requestAuthentication()
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (data == null) {
            Crashlytics.log(Log.ERROR, TAG, "data is null ??? why ????")
            return
        }
        if (requestCode == AuthorizationActivity.REQUEST_AUTHORIZATION) {

            val auth = AuthUtils.activityResultAsAuthentication(requestCode, resultCode, data)
            if (auth != null) {
                authManager!!.onAuthentication(auth)

                when (delete) {
                    Mode.DELETE -> deleteSysten(auth.accessToken!!)
                    Mode.SYNC -> syncWithAv(auth.accessToken!!)
                    Mode.UPDATE -> updateAllSystem(auth.accessToken!!)
                }
            }
        } else if (requestCode == CONFIGURE) {
            if (resultCode == Activity.RESULT_OK) {
                this.delete = Mode.SYNC
                val position = data.getIntExtra(POS, -1)
                //set current and start synchronization

                objectsManager!!.setSavedPosition(position)

                if (checkCredentials()) {
                    val auth = authManager!!.authentication
                    if (!auth!!.isExpired) {
                        syncWithAv(auth.accessToken!!)
                    } else {
                        this.delete = Mode.SYNC
                        requestAuthentication()
                    }
                }
            }
            MainActivity.instance.loadMenu(false)
            reloadMenu()
            objectConfigure.invalidate()

        }
    }

    internal fun delete() {
        if (checkCredentials()) {
            val auth = authManager!!.authentication
            if (!auth!!.isExpired) {
                deleteSysten(auth.accessToken!!)
            } else {
                this.delete = Mode.DELETE
                requestAuthentication()
            }
        }
    }

    private fun updateAllSystem(token: String) {

        if (current >= objectsManager!!.objects.size) return

        objectsManager!!.setSavedPosition(current)

        val prefs = PreferenceUtils.getAvPhonePrefs(activity)

        val display = this

        val updateTask = taskFactory!!.updateTask(prefs.serverHost!!, token)

        val updateParams = UpdateParams()
        updateParams.deviceId = DeviceInfo.getUniqueId(activity)
        updateParams.imei = DeviceInfo.getIMEI(activity)
        updateParams.deviceName = DeviceInfo.deviceName
        updateParams.iccid = DeviceInfo.getICCID(activity)
        updateParams.mqttPassword = prefs.password
        updateParams.customData = PreferenceUtils.getCustomDataLabels(activity)

        updateTask.execute(updateParams)

        updateTask.addProgressListener { result ->

            updateTask.showResult(result, objectsManager!!.savedObjectName!!, display, activity)
            // Update next system
            current++
            updateAllSystem(token)


        }


    }

    private fun deleteSysten(token: String) {
        val prefs = PreferenceUtils.getAvPhonePrefs(activity)

        val display = this

        val deleteTask = taskFactory!!.deleteSystemTak(prefs.serverHost!!, token)
        deleteTask.execute()

        deleteTask.addProgressListener({ result ->
            if (delete == Mode.DELETE) {
                objectsManager!!.removeSavedObject()
            }

            deleteTask.showResult(result, display, activity)
            MainActivity.instance.loadMenu(false)
            reloadMenu()
            objectConfigure.invalidate()

        })


    }

    private fun syncWithAv(token: String) {

        val prefs = PreferenceUtils.getAvPhonePrefs(activity)

        val display = this

        val syncTask = taskFactory!!.syncAvTask(prefs.serverHost!!, token)

        val syncParams = SyncWithAvParams()
        syncParams.deviceId = DeviceInfo.getUniqueId(activity)
        syncParams.imei = DeviceInfo.getIMEI(activity)
        syncParams.deviceName = DeviceInfo.deviceName
        syncParams.iccid = DeviceInfo.getICCID(activity)
        syncParams.mqttPassword = prefs.password
        syncParams.customData = PreferenceUtils.getCustomDataLabels(activity)

        syncTask.execute(syncParams)

        syncTask.addProgressListener { result ->
            if (delete == Mode.DELETE) {
                objectsManager!!.removeSavedObject()
            }

            syncTask.showResult(result, display, activity)
            MainActivity.instance.loadMenu(false)

            if (!result.isError) {
                syncListener!!.invoke(result)
            }
        }

    }

    companion object {
        var INDEX = "index"
        var CONFIGURE = 0
        var POS = "position"
        @SuppressLint("StaticFieldLeak")
        var instance: ConfigureFragment? = null
        private val TAG = ConfigureFragment::class.simpleName
    }

}
